DPL: finally use concepts to separate the make method

ktf opened the pull request:
In the past, substituting SFINAE with constexpr made sense because
understanding what was going on in case of failure was extremely complicated.

Now that we have constrains and cocepts, it makes sense to split again the
method since the diagnostics are good and the compiler is limited to optimise
the code in question, without actually having to instanciate all the templates.

alibuild commented:
Error while checking build/O2/fullCI for 96537e5b8b3eac1967d4f377b1c8903a3bf1edde at 2024-10-18 13:35:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
/sw/SOURCES/O2/13611-slc8_x86-64/0/GPU/Workflow/src/GPUWorkflowSpec.cxx:714:55: error: 'struct o2::framework::DataAllocator::UninitializedVector<char>' has no member named 'data'
/sw/SOURCES/O2/13611-slc8_x86-64/0/GPU/Workflow/src/GPUWorkflowSpec.cxx:718:63: error: 'struct o2::framework::DataAllocator::UninitializedVector<char>' has no member named 'data'
/sw/SOURCES/O2/13611-slc8_x86-64/0/GPU/Workflow/src/GPUWorkflowSpec.cxx:719:43: error: 'struct o2::framework::DataAllocator::UninitializedVector<char>' has no member named 'size'
/sw/SOURCES/O2/13611-slc8_x86-64/0/GPU/Workflow/src/GPUWorkflowSpec.cxx:729:29: error: 'struct o2::framework::DataAllocator::UninitializedVector<char>' has no member named 'size'
/sw/SOURCES/O2/13611-slc8_x86-64/0/GPU/Workflow/src/GPUWorkflowSpec.cxx:732:25: error: 'struct o2::framework::DataAllocator::UninitializedVector<char>' has no member named 'resize'
/sw/SOURCES/O2/13611-slc8_x86-64/0/GPU/Workflow/src/GPUWorkflowSpec.cxx:733:37: error: 'struct o2::framework::DataAllocator::UninitializedVector<char>' has no member named 'data'
/sw/SOURCES/O2/13611-slc8_x86-64/0/GPU/Workflow/src/GPUWorkflowSpec.cxx:706:31: error: no matching function for call to 'std::optional<std::reference_wrapper<o2::framework::DataAllocator::UninitializedVector<char> > >::emplace(gsl::span<o2::framework::DataAllocator::UninitializedVector<char>, 18446744073709551615>&)'
/sw/slc8_x86-64/GCC-Toolchain/v13.2.0-alice1-6/include/c++/13.2.0/type_traits:2610:11: error: no type named 'type' in 'struct std::enable_if<false, std::reference_wrapper<o2::framework::DataAllocator::UninitializedVector<char> >&>'
/sw/SOURCES/O2/13611-slc8_x86-64/0/GPU/Workflow/src/GPUWorkflowSpec.cxx:717:29: error: no matching function for call to 'std::optional<std::reference_wrapper<o2::framework::DataAllocator::UninitializedVector<char> > >::emplace(gsl::span<o2::framework::DataAllocator::UninitializedVector<char>, 18446744073709551615>&)'
/sw/SOURCES/O2/13611-slc8_x86-64/0/GPU/Workflow/src/GPUWorkflowSpec.cxx:706:31: error: no matching function for call to 'std::optional<std::reference_wrapper<o2::framework::DataAllocator::UninitializedVector<char> > >::emplace(gsl::span<o2::framework::DataAllocator::UninitializedVector<char>, 18446744073709551615>&)'
/sw/SOURCES/O2/13611-slc8_x86-64/0/GPU/Workflow/src/GPUWorkflowSpec.cxx:717:29: error: no matching function for call to 'std::optional<std::reference_wrapper<o2::framework::DataAllocator::UninitializedVector<char> > >::emplace(gsl::span<o2::framework::DataAllocator::UninitializedVector<char>, 18446744073709551615>&)'
/sw/SOURCES/O2/13611-slc8_x86-64/0/GPU/Workflow/src/GPUWorkflowSpec.cxx:869:39: error: 'struct o2::framework::DataAllocator::UninitializedVector<char>' has no member named 'resize'
/sw/SOURCES/O2/13611-slc8_x86-64/0/GPU/Workflow/src/GPUWorkflowSpec.cxx:870:65: error: 'struct o2::framework::DataAllocator::UninitializedVector<char>' has no member named 'data'
/sw/SOURCES/O2/13611-slc8_x86-64/0/GPU/Workflow/src/GPUWorkflowSpec.cxx:906:106: error: 'struct o2::framework::DataAllocator::UninitializedVector<char>' has no member named 'data'
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13611/96537e5b8b3eac1967d4f377b1c8903a3bf1edde/build_O2_fullCI/pretty.html).

ktf committed:
DPL: finally use concepts to separate the make method

In the past, substituting SFINAE with constexpr made sense because
understanding what was going on in case of failure was extremely complicated.

Now that we have constrains and cocepts, it makes sense to split again the
method since the diagnostics are good and the compiler is limited to optimise
the code in question, without actually having to instanciate all the templates.
